---
source: Rmd
title: "Introduction to Bioconductor"
teaching: XX
exercises: XX
questions:
- "What does the Bioconductor project comprise?"
- "How does the Bioconductor project relate to the CRAN repository?"
objectives:
- "Describe the Bioconductor project globally."
- 'Gain a global view of the R <i class="fab fa-r-project"></i> ecosystem.'
keypoints:
- "Bioconductor is a repository of R packages, but also much more than that."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("01-")
```

```{r, include=FALSE}
options(htmltools.dir.version = FALSE)
library(RefManageR)
bib <- ReadBib("../bibliography.bib")
BibOptions(
  longnamesfirst = FALSE,
  check.entries = FALSE,
  bib.style = "authoryear",
  cite.style = "authoryear",
  max.names = 1,
  style = "html",
  hyperlink = "to.bib",
  dashed = TRUE)
```

> ## Contribute!
> 
> This episode is in particular need of interactive elements.
> See <https://carpentries.github.io/curriculum-development/designing-challenges.html#different-types-of-challenges> for different types of challenges, e.g.:
> 
> - Multiple Choice Questions
> - Fill-in-the-blank
>
{: .callout}

# What is Bioconductor?

## A scientific project

The original publication describes the aims and methods of the project at its inception `r Citet(bib, "Gentleman2004")`.

`r Citet(bib, "Huber2015")` illustrates the progression of the project, including descriptions of core infrastructure and case studies, from the perspective of both users and developers.

`r Citet(bib, "Amezquita2020")` reviewed further developments of the project in the wake of single-cell genomics technologies.

Many more publications and book chapters cite the Bioconductor project, with recent example listed on the [Bioconductor website][bioc-publications].


## A package repository

Undoubtedly what the Bioconductor project is best known for, the repository has grown over time to host thousands of packages.

Packages are broadly divided in four major categories:

- software
- annotation data
- experiment data
- workflows

Software packages themselves can be subdivided into packages that provide infrastructure (i.e., classes) to store and access data,
and packages that provide methodological tools to process data stored in those data structures.
This separation of structure and analysis is at the core of the Bioconductor project,
encouraging developers of new methodological software packages to thoughtfully re-use existing data containers where possible,
and reducing the cognitive burden imposed on users who can more easily experiment with alternative workflows without the need to learn and convert between different data structures.

Annotation packages provide self-contained databases of diverse genomic annotations (e.g., gene identifiers, biological pathways).
Different collections of annotation packages can be found in the Bioconductor project.
They are identifiable by their respective naming pattern, and the information that they contain.
For instance, the so-called `OrgDb` packages (e.g., the `r BiocStyle::Biocpkg("org.Hs.eg.db")` package) provide information mapping different types of gene identifiers and pathway databases;
the so-called `EnsDb` (e.g., `r BiocStyle::Biocpkg("EnsDb.Hsapiens.v86")`) packages encapsulate individual versions of the Ensembl annotations in Bioconductor packages;
and the so-called `TxDb` packages (e.g., `r BiocStyle::Biocpkg("TxDb.Hsapiens.UCSC.hg38.knownGene")`) encapsulate individual versions UCSC gene annotation tables.

Experiment data package provide self-contained datasets that are often used by software package developers to demonstrate the use of their package on well-known standard datasets in their [package vignettes][package-vignette].

Finally, workflow packages ...

[BiocViews][biocviews-software] are a predefined - yet, evolving - set of labels used to categorize all the packages available in the Bioconductor project.


## Online communication channels

### A support site

The Bioconductor [support site][bioc-support-site] provides a platform where users and developers can communicate freely (following the Bioconductor [Code of Conduct][bioc-code-of-conduct]) to discuss issues on a range of subjects, ranging from packages to conceptual questions about best practices.


### A Slack workspace

The Bioconductor [Slack workspace][bioc-slack] is an open space that all community members are welcome to join (for free) and use for rapid interactions.
Currently, the "Pro" pricing plan kindly supported by core funding provides:

- Unlimited message archive
- Unlimited apps
- Group video calls with screen sharing
- Work securely with other organizations using Slack Connect

A wide range of channels have been created to discuss a range of subjects, and community members can freely join the discussion on those channels of create new ones to discuss new subjects.

Important announcements are posted on the `#general` channel.

> ## Note
>
> Users are encouraged to use the Bioconductor [support site][bioc-support-site] to raise issues that relevant to the wider community.
> The Slack workspace is often most useful for live discussions, and widely subscribed channels (e.g. `#general`) should be used with moderation.
>
{: .callout}


### A mailing list for developers

The <bioc-devel@r-project.org> mailing list is used for communication between package developers, and announcements from the Biocondutor core team.


## Conferences, courses and workshops

The Bioconductor community regularly organizes a number of events throughout the year and across the world.
For example:

- The annual BioC summer conference in North America
- Regional conference in winter (e.g. BioC Europe, BioC Asia)
- Summer schools (e.g., CSAMA)
- Online meetings open to all community members (e.g., Bioconductor Developers Forum)

[Course materials][bioc-course-materials] are regularly uploaded on the Bioconductor website following each of those events.


## A scientific and technical community

> ## Contribute!
>
> - Add some details to the role of each panel below.
>
{: .callout}

- [Scientific Advisory Board][bioc-scientific-advisory-board]
- [Technical Advisory Board][bioc-technical-advisory-board]
- [Community Advisory Board][bioc-community-advisory-board]


# References

```{r, results='asis', echo=FALSE}
PrintBibliography(bib)
```

[biocviews-software]: https://www.bioconductor.org/packages/release/BiocViews.html#___Software
[bioc-code-of-conduct]: https://www.bioconductor.org/about/code-of-conduct/
[bioc-slack]: https://bioc-community.herokuapp.com/
[bioc-support-site]: https://support.bioconductor.org/
[bioc-course-materials]: https://bioconductor.org/help/course-materials/
[bioc-publications]: https://www.bioconductor.org/help/publications/
[bioc-scientific-advisory-board]: https://bioconductor.org/about/scientific-advisory-board/
[bioc-technical-advisory-board]: https://bioconductor.org/about/technical-advisory-board/
[bioc-community-advisory-board]: https://bioconductor.org/about/community-advisory-board/
[package-vignette]: ../reference.html#vignette
