---
source: Rmd
title: "Introduction to Bioconductor"
teaching: XX
exercises: XX
questions:
- "What does the Bioconductor project comprise?"
- "How does the Bioconductor project relate to the CRAN repository?"
objectives:
- "Describe the Bioconductor project globally."
- 'Gain a global view of the R <i class="fab fa-r-project"></i> ecosystem.'
keypoints:
- "Bioconductor is a repository of R packages, but also much more than that."
- "The Bioconductor project regularly releases versions of the entire package repository, promoting traceability and reproducibility."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("01-")
```

```{r, include=FALSE}
options(htmltools.dir.version = FALSE)
library(RefManageR)
bib <- ReadBib("../bibliography.bib")
BibOptions(
  longnamesfirst = FALSE,
  check.entries = FALSE,
  bib.style = "authoryear",
  cite.style = "authoryear",
  max.names = 1,
  style = "html",
  hyperlink = "to.bib",
  dashed = TRUE)
```

> ## Contribute!
> 
> This episode is in particular need of interactive elements.
> See <https://carpentries.github.io/curriculum-development/designing-challenges.html#different-types-of-challenges> for different types of challenges, e.g.:
> 
> - Multiple Choice Questions
> - Fill-in-the-blank
>
{: .callout}

# What is Bioconductor?

## A brief history of Bioconductor

The Bioconductor project was start in the Fall of 2001, as an initiative for the collaborative creation of extensible software for computational biology and bioinformatics `r Citep(bib, "Gentleman2004")`.
From the very start, the core focus of the project was to develop tools for the statistical analysis and comprehension of large datasets and technological artifacts in rigorously and robustly designed experiments.
Beyond statistical analyses, the interpretation of statistical results is supported by packages providing biological context, visualization, and reproducibility.

Over the years, software packages contributed to the Bioconductor project have reflected the evolution and emergence of several high-throughput technologies, from microarrays to single-cell genomics, through many variations of sequencing experiments (e.g., RNA-seq, ChIP-seq, DNA-seq), analyses (e.g., sequence variation, copy number variation, single nucleotide polymorphisms), and data modalities (e.g., flow cytometry, proteomics, microscropy and image analysis).

Crucially, the project has not only released software packages implementing novel statitical tests and methodologies, but also produced a diverse range of packages types granting access to databases of molecular annotations and experimental datasets.

The Bioconductor project culminates at an annual conference in North America in the summer, while regional conferences offer great opportunities for networking in Europe, Asia, and North America.
The project is commited to promote a diverse and inclusive community, including a [Code of Conduct][bioc-code-of-conduct] enforced by a Code of Conduct committee.

> ## Contribute!
> 
> It would be great to insert an illustration of the timeline of the Bioconductor project, including major milestones, e.g.
> - First commit
> - First completed package
> - First official release
> - First release of the main website
> - First summer conference
> - First regional conference (e.g., EuroBioC, BioCAsia)
> - First conference workshop
> - First appearance of various standard classes (e.g., `ExpressionSet`, `GRanges`, `DNAString`, `SummarizedExperiment`, `SingleCellExperiment`)
> - First annotation package
> - First release of the support website
> - Creation of the Slack workspace
> - Creation of each advisory board
>
> The timeline should not aim to be comprehensive and overloaded.
> Rather, it should provide a broad brush overview of major milestones that illustrate the evolution of the project over time.
{: .callout}

## A scientific project

The original publication describes the aims and methods of the project at its inception `r Citet(bib, "Gentleman2004")`.

`r Citet(bib, "Huber2015")` illustrates the progression of the project, including descriptions of core infrastructure and case studies, from the perspective of both users and developers.

`r Citet(bib, "Amezquita2020")` reviewed further developments of the project in the wake of single-cell genomics technologies.

Many more publications and book chapters cite the Bioconductor project, with recent example listed on the [Bioconductor website][bioc-publications].

## A package repository

### Overview and relationship to CRAN

Undoubtedly, software packages are the best-known aspect of the Bioconductor project,.
Since its inception in 2001, the repository has grown over time to host thousands of packages.

The Bioconductor project has extended the preexisting CRAN repository -- much larger and general-purpose in scope -- to comprise R packages primarily catering for bioinformatics and computational biology analyses.

### The Bioconductor release cycle

Furthermore, the Bioconductor project also extended the packaging infrastructure of the CRAN repository to better support the deployment and management of packages at the user level `r Citet(bib, "Gentleman2004")`.
In particular, the Bioconductor projects features a 6-month release cycle (typically around April and October), which sees a snapshot of the current version of all packages in the Bioconductor repository earmarked for a specific version of R.
R itself is released on an annual basis (typically around April), meaning that for each release of R, two compatible releases of Bioconductor packages are available.

As such, Bioconductor package developers are required to always use the version of R that will be associated with the next release of the Bioconductor project.
This means using the development version of R between October and April, and the release version of R between April and October.

Crucially, the strict Bioconductor release cycle prevents users from installing temporally distant versions of packages that were very likely never tested together.
This practice reflects the development cycle of packages of both CRAN and Bioconductor, where contemporaneous packages are regularly tested by automated systems to ensure that the latest software updates in package dependencies do not break downstream packages, or prompts those package maintainers to update their own software as a consequence.

Prior to each Bioconductor release, packages that do not pass the requires suites of automated tests are deprecated and subsequently removed from the repository.
This ensures that each Bioconductor release provides a suite of packages that are mutually compatible, traceable, and guaranteed to function for the associated version of R.

> ## Contribute!
> 
> It would be great to insert an illustration of the Bioconductor release cycle, to accompany this blob of text.
{: .callout}

### Package types

Packages are broadly divided in four major categories:

- software
- annotation data
- experiment data
- workflows

[Software packages][glossary-software-package] themselves can be subdivided into packages that provide infrastructure (i.e., classes) to store and access data,
and packages that provide methodological tools to process data stored in those data structures.
This separation of structure and analysis is at the core of the Bioconductor project,
encouraging developers of new methodological software packages to thoughtfully re-use existing data containers where possible,
and reducing the cognitive burden imposed on users who can more easily experiment with alternative workflows without the need to learn and convert between different data structures.

[Annotation data packages][glossary-annotation-package] provide self-contained databases of diverse genomic annotations (e.g., gene identifiers, biological pathways).
Different collections of annotation packages can be found in the Bioconductor project.
They are identifiable by their respective naming pattern, and the information that they contain.
For instance, the so-called `OrgDb` packages (e.g., the `r BiocStyle::Biocpkg("org.Hs.eg.db")` package) provide information mapping different types of gene identifiers and pathway databases;
the so-called `EnsDb` (e.g., `r BiocStyle::Biocpkg("EnsDb.Hsapiens.v86")`) packages encapsulate individual versions of the Ensembl annotations in Bioconductor packages;
and the so-called `TxDb` packages (e.g., `r BiocStyle::Biocpkg("TxDb.Hsapiens.UCSC.hg38.knownGene")`) encapsulate individual versions UCSC gene annotation tables.

[Experiment data packages][glossary-experiment-package] provide self-contained datasets that are often used by software package developers to demonstrate the use of their package on well-known standard datasets in their [package vignettes][glossary-vignette].

Finally, [workflow packages][glossary-workflow-package] exclusively provide collections of vignettes that demonstrate the combined usage of several other packages as a coherent workflow, but do not provide any new source code or functionality themselves.

> ## Challenge: The Bioconductor website
>
> The Bioconductor website is accessible at <https://bioconductor.org/>.
> 
> Browse the website to find information answering the following questions:
>
> 1. How many packages does the current release of the Bioconductor project include?
> 2. How many packages of each type does this number include?
>
> > ## Solution
> >
> > The following solution includes numbers that were valid at the time of writing (Bioconductor release 3.13);
> > numbers will inevitably be different for future releases of the Bioconductor project.
> >
> > 1. On the page <https://bioconductor.org/>, in the section "Install", we can read:
> > > Discover 2042 software packages available in Bioconductor release 3.13.
> >
> > 2. On the page <https://bioconductor.org/>, in the section "News", click on the link that reads "Bioconductor Bioc `X.Y` Released" (`X.Y` being the version of the current Bioconductor release when you go through this exercise yourself).
> >    On the linked page, we can read:
> > > We are pleased to announce Bioconductor 3.13, consisting of 2042 software packages, 406 experiment data packages, 965 annotation packages, and 29 workflows.
> > > 
> > > There are 133 new software packages, 22 new data experiment packages, 7 new annotation packages, 1 new workflow, no new books, and many updates and improvements to existing packages; Bioconductor 3.13 is compatible with R 4.1.0, and is supported on Linux, 32- and 64-bit Windows, and macOS 10.14.6 Mojave or higher. This release will include an updated Bioconductor Docker containers.
> >
> {: .solution}
{: .challenge}

### Package classification using biocViews

The Bioconductor project uses [biocViews][glossary-biocviews], a set of terms from a controlled vocabulary, to classify Bioconductor packages and facilitate their discovery by thematic search on the [Bioconductor website][biocviews-site].

Each Bioconductor package is tagged with a small set of terms chosen from the available controlled vocabulary, to describe the type and functionality of the package.
Terms are initially selected by the package authors, and subsequently refined during package review or updates to the controlled vocabulary.

For instance, the `r BiocStyle::Biocpkg("BiocPkgTools")` package can be used to browse packages under different [biocViews][glossary-biocviews] `r Citep(bib, "BiocPkgTools2019")`.

> ## Going further
> 
> The Bioconductor package `r BiocStyle::Biocpkg("biocViews")` is used to support and manage the infrastructure of the controlled vocabulary.
> It can also be used to programmatically inspect and subset the list of terms available using their relationship as a graph.
{: .callout}

### Packages interoperability

At the core of the Bioconductor philosophy is the notion of interoperability.
Namely, the capacity of packages to operate on the same data structures.
Importantly, interoperability benefits both users and developers.

Users can more easily write arbitrarily complex workflows that combine multiple packages.
With packages operating on the same data structure, users can maximize their attention to the practical steps of their workflow, and minimize time spent in often complex and error-prone conversions between different data structures specific to each package.
Comparative benchmarks are also easier to implement and can be evaluated more fairly
when competing software packages offering similar functionality operate on input and outputs stored in the same data structures.

Similarly, developers of new packages can focus on the implementation of novel functionality, borrowing existing data structures that offer robust and trusted infrastructure for storage, verification, and indexing of information.

Ultimately, the figure below illustrates how many different Bioconductor packages - as well as base R packages - can be combine for carry out a diverse range of analyses, from importing sequencing data into an R session, to the annotation, integration and visualization of data and results.

![Sequencing Ecosystem](../fig/bioc-sequencing-ecosystem.png)

## Conferences, courses and workshops

The Bioconductor community regularly organizes a number of events throughout the year and across the world. For example:

- The annual BioC summer conference in North America
- Regional conference in winter (e.g. BioC Europe, BioC Asia)
- Summer schools (e.g., CSAMA)
- Online meetings open to all community members (e.g., Bioconductor Developers Forum)

Course materials are regularly uploaded on the [Bioconductor website][bioc-course-materials] following each of those events.
In particular, [online books][bioc-books] are being developed and maintained by community members.

The Bioconductor [YouTube channel][bioc-youtube] is used to publish video recordings of conference presentations including talks and workshops, as well as editions of the regular Bioconductor developers forum (link needed).

> ## Contribute!
>
> It could be great to illustrate a typical cycle of conferences over a year, e.g.
> - BioC conference in North America around late July
> - EuroBioC conference in Europe around December
> - BioCAsia conference in Asia around November
{: .callout}

## Online communication channels

### A support site

The Bioconductor [support site][bioc-support-site] provides a platform where users and developers can communicate freely (following the Bioconductor [Code of Conduct][bioc-code-of-conduct]) to discuss issues on a range of subjects, ranging from packages to conceptual questions about best practices.

### A Slack workspace

The Bioconductor [Slack workspace][bioc-slack] is an open space that all community members are welcome to join (for free) and use for rapid interactions.
Currently, the "Pro" pricing plan kindly supported by core funding provides:

- Unlimited message archive
- Unlimited apps
- Group video calls with screen sharing
- Work securely with other organizations using Slack Connect

A wide range of channels have been created to discuss a range of subjects, and community members can freely join the discussion on those channels of create new ones to discuss new subjects.

Important announcements are posted on the `#general` channel.

> ## Note
>
> Users are encouraged to use the Bioconductor [support site][bioc-support-site] to raise issues that relevant to the wider community.
> The Slack workspace is often most useful for live discussions, and widely subscribed channels (e.g. `#general`) should be used with moderation.
>
{: .callout}

### A mailing list for developers

The <bioc-devel@r-project.org> mailing list is used for communication between package developers, and announcements from the Biocondutor core team.

## A scientific and technical community

> ## Contribute!
>
> - Add some details to the role of each panel below.
>
{: .callout}

- [Scientific Advisory Board][bioc-scientific-advisory-board]
- [Technical Advisory Board][bioc-technical-advisory-board]
- [Community Advisory Board][bioc-community-advisory-board]

# References

```{r, results='asis', echo=FALSE}
PrintBibliography(bib)
```

[biocviews-site]: https://www.bioconductor.org/packages/release/BiocViews.html
[bioc-code-of-conduct]: https://www.bioconductor.org/about/code-of-conduct/
[bioc-slack]: https://bioc-community.herokuapp.com/
[bioc-support-site]: https://support.bioconductor.org/
[bioc-course-materials]: https://bioconductor.org/help/course-materials/
[bioc-publications]: https://www.bioconductor.org/help/publications/
[bioc-scientific-advisory-board]: https://bioconductor.org/about/scientific-advisory-board/
[bioc-technical-advisory-board]: https://bioconductor.org/about/technical-advisory-board/
[bioc-community-advisory-board]: https://bioconductor.org/about/community-advisory-board/
[bioc-youtube]: https://www.youtube.com/user/bioconductor
[bioc-books]: http://www.bioconductor.org/books/release/
[bioc-code-of-conduct]: https://www.bioconductor.org/about/code-of-conduct/
[glossary-vignette]: ../reference.html#vignette
[glossary-software-package]: ../reference.html#software-package
[glossary-annotation-package]: ../reference.html#annotationdata-package
[glossary-experiment-package]: ../reference.html#experimentdata-package
[glossary-workflow-package]: ../reference.html#workflow-package
[glossary-biocviews]: ../reference.html#biocviews
