---
source: Rmd
title: The SummarizedExperiment class
teaching: XX
exercises: XX
---

---

```{r, echo=FALSE, purl=FALSE, message=FALSE}
```

::::::::::::::::::::::::::::::::::::::: objectives

- Describe how both experimental data and metadata can be stored in a single object.
- Explain why this is crucial to keep data and metadata synchronised throughout analyses.

::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: questions

- How is information organized in SummarizedExperiment objects?
- How can that information be added, edited, and accessed?

::::::::::::::::::::::::::::::::::::::::::::::::::

```{r, include=FALSE}
```

```{r, include=FALSE}
options(htmltools.dir.version = FALSE)
library(RefManageR)
library(bibtex)
bib <- ReadBib("files/bibliography.bib")
```

## Install packages

Before we can proceed into the following sections, we install some Bioconductor
packages that we will need.
First, we check that the `r BiocStyle::Biocpkg("BiocManager")` package is
installed before trying to use it; otherwise we install it.
Then we use the `BiocManager::install()` function to install the necessary
packages.

```{r, message=FALSE, warning=FALSE, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c("SummarizedExperiment"))
```

## Motivation

Experiments are multifaceted data sets typically composed of at least two key
pieces of information necessary for any analysis:

- Assay data, typically a matrix representing measurements of a set of features
  in a set of samples
  (e.g., RNA-sequencing).
- Sample metadata, typically a `data.frame` of metadata representing information
  about samples
  (e.g., treatment group).

All those pieces of information must be kept synchronised -- same samples,
same order -- for downstream analyses to accurately process the information
and produce reliable results.

It is also very common for analytical workflows to analyse subsets of samples or
identify outliers that need to be removed to allow for more accurate downstream
analyses.
In such cases, all aspects of the experiments must be subsetted to the same set
of samples -- in the same order -- to preserve consistency in the data set
and correct results.

The `SummarizedExperiment` -- implemented in the
`r BiocStyle::Biocpkg("SummarizedExperiment")` package -- provides a container
that accommodates those essential aspects of individual experiments into a
single object coordinates data and metadata during subsetting and reordering
operations.

## Class structure

`SummarizedExperiment` is a matrix-like container where rows represent features of
interest (e.g. genes, transcripts, exons, etc.) and columns represent samples.

The objects can contain one or more assays, each represented by a matrix-like
object, as long as they be of the same dimensions.

Information about the features is stored in a `DataFrame` object,
nested within the `SummarizedExperiment` object,
and accessible using the function `rowData()`.
Each row of the `DataFrame` provides information on the feature in the
corresponding row of the SummarizedExperiment object.
That information may include annotations independent of the experiment
(e.g., gene identifier) as well as quality control metrics computed from
assay data during workflows.

Similarly, information about the samples is stored in another `DataFrame` object,
also nested within the `SummarizedExperiment` object,
and accessible using the function `colData()`.

The following graphic displays the class geometry and highlights the
vertical (column) and horizontal (row) relationships.
It was obtained from the vignette of the
`r BiocStyle::Biocpkg("SummarizedExperiment")` package.

![](fig/summarizedexperiment.svg){alt='Schematic representation of the SummarizedExperiment class.'}

## Creating a SummarizedExperiment object

Let us first load the package:

```{r, message=FALSE, warning=FALSE}
library(SummarizedExperiment)
```

Then, let us import assay data from a file that we downloaded during the lesson
setup.

The file is a simple text file in which
the first column contains made-up feature identifiers
and all other columns contain simulated data for made-up samples.
As such, we can use the base R function `read.csv` to parse the file
into a `data.frame` object.

In the example below, we indicate that the row names can be found in the first
column, so that the function immediately sets the row names accordingly
in the output object.
Hadn't we specified it, the function would have parsed it as a regular column
and left the row names to the default integer indexing.

```{r}
count_data <- read.csv("data/counts.csv", row.names = 1)
count_data
```

[biocviews]: https://www.bioconductor.org/packages/release/BiocViews.html
[biomart-ensembl]: https://www.ensembl.org/biomart/martview
